### PR TITLE
add option of skipping test case when first failure occured (#9)

### PR DESCRIPTION
**Pull request by reAsOn2010**
used for ACM scoring rule...

### CODE REVIEW

File: src/ljudge.cc [hunk 1]
```diff
@@ -1413,6 +1417,11 @@ static Options parse_cli_options(int argc, const char *argv[]) {
       REQUIRE_NARGV(1);
       options.nthread = NEXT_NUMBER_ARG;
 #endif
+    } else if (option == "skip-on-first-failure") {
+      if (options.nthread != 1) {
```
Comment (quark-zju) [inline src/ljudge.cc]: Perhaps `nthread` can be automatically set to 1 in this case so `--skip-on-first-failure` works without specifying threads explicitly.

**Comment by reAsOn2010** [comment]
I saw the code already merged into master, so close.

File: src/ljudge.cc [hunk 2]
```diff
@@ -2124,12 +2133,22 @@ static j::value run_testcases(const Options& opts) {
...
-    results[i] = j::value(testcase_result);
+    if (opts.skip_on_first_failure && should_skip) {
+      j::object skipped_result;
+      skipped_result["result"] = j::value(TestcaseResult::SKIPPED);
+      results[i] = j::value(skipped_result);
+    } else {
+      j::object testcase_result = run_testcase(opts.etc_dir, opts.cache_dir, opts.user_code_path, opts.checker_code_path, opts.envs, opts.cases[i], opts.skip_checker, opts.keep_stdout, opts.keep_stderr);
+      results[i] = j::value(testcase_result);
+      if (testcase_result["result"].to_str() != TestcaseResult::ACCEPTED) {
+        should_skip = true;
```
Comment (quark-zju) [inline src/ljudge.cc:2148]: This seems a bit tricky for OpenMP. Although it compiles, I think it's better to make the openmp loop explicitly not mutate a variable outside the loop body.

File: src/ljudge.cc [hunk 3]
```diff
@@ -108,6 +108,7 @@ namespace TestcaseResult {
   const string FLOAT_POINT_EXCEPTION = "FLOAT_POINT_EXCEPTION";
   const string SEGMENTATION_FAULT = "SEGMENTATION_FAULT";
   const string WRONG_ANSWER = "WRONG_ANSWER";
+  const string SKIPPED = "SKIPPED";
```
Comment (quark-zju) [inline src/ljudge.cc:111]: Lines here are generated via [cog](https://nedbatchelder.com/code/cog/). I have updated the cog comment.

**Review by quark-zju — COMMENTED** [review]
The change was pushed as bece9a4051bfe67fefa13e58c2e2f2981ac7bf2c. Thanks!

I've made some minor edits.

File: src/ljudge.cc (hunk 2 above)
Comment (reAsOn2010) [reply src/ljudge.cc:2148]: I see, seems better.

**Pull request closed without merging**